### DISCUSSION TITLE
Terminology

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#360] [General]
This is a thread to discuss Dagger’s terminology (the choice of specific words to designate specific things), our experience with it, suggestions for changing it, etc.

Dagger’s terminology is still changing rapidly. Product terminology is part of product design, and the only way to find the right design is to try lots of bad ones. So, we’ve been trying lots of words to describe the important parts of Dagger. As we use the product, we get a better idea of which words work and which don’t.

Note: this thread is not specific to any particular version of the terminology.

## Current terminology

Here is a sample of the current Dagger terminology. I will keep this section updated to match the reality of the code & docs.

- *Deployment*: a continuously running workflow delivering a specific application in a specific way.
- *Deployment Plan*: the source code of the deployment, written in Cue.
- *Relay*: a standalone software component assigned to a single node in the deployment plan.
- *Input*: data supplied to a deployment by the user
- *Version*: a snapshot in time of the state of a deployment.

**Comment by shykes** [comment]
After a few weeks of feedback and hands-on practice, I am leaning towards the following changes:

> * _Deployment_: a continuously running workflow delivering a specific application in a specific way.

Rename to *Environment*. Rationale:

* It comes up all the time in natural conversation. No matter what the terminology says, I end up using the word “environment” to explain it.
* Solves the ambiguity of the word “deployment” which could be a) the action of deploying once, or b) the result of one or more deployment actions. No such ambiguity with environment.

> * _Deployment Plan_: the source code of the deployment, written in Cue.

Rename to *Workflow*. Rationale:

* Comes up in natural conversation, regardless of terminology.
* “deployment plan” no longer works if we rename “deployment”
* a single word is better than 2 words. If it’s a fundamental concept, it shouldn’t need a qualifier.
* it’s a good fit for something that is code but not state (“these 5 environments are all executing the same workflow with different inputs”).

**Reply by aluzzardi** [reply]
SGTM

Do you want to go ahead with the renaming or still need some time to see how it feels?

**Reply by shykes** [reply]
> Do you want to go ahead with the renaming or still need some time to see how it feels?

100% conviction on renaming “deployment” to “environment”. In great part because of your feedback on the ambiguous meaning of deployment.

80% conviction on renaming “deployment plan” to “workflow”. We could just rename it to “plan”. But when not combined with “deployment”, it’s not as strong IMO.

Either way I’d like to make a change today, the earlier the better.

**Reply by aluzzardi** [reply]
Sounds good.

Personally I really like both changes.

**Reply by samalba** [reply]
Light preference on `plan` rather than `workflow` for me. But fine either way.

**Reply by icecrime** [reply]
I’ve mentioned it before, I think `environment` is a perfect fit for the first concept 👍

I’m not a fan of `workflow`, which is also pretty overloaded and sounds like something sequential (which dagger isn’t). I’ve always like Ansible’s terminology of `playbook`, and more generally something like `plan` or `recipe` sounds good to me too.

*PS: I feel bad for bike-shedding on naming without taking the time to participate on deeper topics.*

**Reply by shykes** [reply]
Don’t feel bad. I don’t consider this bikeshedding, more like structuralwalling :) You’re a busy guy and any contribution is appreciated.

Personally I am still hesitating between `workflow` and `plan`. @samalba and @icecrime : do you think there might be confusion with the concept of `terraform plan` and `pulumi plan`?

**Reply by shykes** [reply]
Update: I opened a PR to replace `deployment + plan` with `environment + plan`. https://github.com/dagger/dagger/pull/368.

 I think “plan” could work, I agree it is less overloaded than “workflow”.

How does everyone feel about possible conflict between the concept of a Dagger plan, with the actions `terraform plan`, `pulumi plan` and (perhaps one day?) and equivalent `dagger plan` ?

**Comment by SvenDowideit** [comment]
the use of relay (as i was reading the docs) for a node in the graph/pipeline feels disjointed to me - it might work better if you explain the etymology behind your use of it?
